File: algorithms/sort/merge-sort/merge-sort.rb
def merge_sort arr
  return arr if arr.size < 2

  middle = (arr.length / 2).floor
  left = arr.take middle
  right = arr.drop middle

  merge merge_sort(left), merge_sort(right)
end

def merge arr1, arr2
  result = []

  while !arr1.empty? && !arr2.empty?
    min = arr1[0] < arr2[0] ? arr1.delete_at(0) : arr2.delete_at(0)
    result << min
  end

  result.concat(!arr1.empty? ? arr1 : arr2)
end

File: algorithms/sort/merge-sort/mergeSort.js
const mergeSort = arr => {
  if (arr.length < 2) return arr

  const middle = Math.floor(arr.length / 2),
        left = arr.slice(0, middle),
        right = arr.slice(middle)

  return merge(mergeSort(left), mergeSort(right))
}

const merge = (arr1, arr2) => {
  let result = []

  while (arr1.length && arr2.length) {
    const min = arr1[0] < arr2[0] ? arr1.splice(0, 1) : arr2.splice(0, 1)
    result = result.concat(min)
  }

  return result.concat(arr1.length ? arr1 : arr2)
}
